pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,260 @@
+
+	c := &controller{
+		queue: queue,
+		getCRD: func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+			return crdInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		getAPIBindingsByBoundResourceUID: func(name string) ([]interface{}, error) {
+			return apiBindingInformer.Informer().GetIndexer().ByIndex(indexers.APIBindingByBoundResourceUID, name)
+		},
+		listAPIBindings: func() ([]*apisv1alpha1.APIBinding, error) {
+			return apiBindingInformer.Lister().List(labels.Everything())
+		},
+		deleteCRD: func(ctx context.Context, name string) error {
+			return crdClusterClient.ApiextensionsV1().CustomResourceDefinitions().Cluster(apibinding.ShadowWorkspaceName).Delete(ctx, name, metav1.DeleteOptions{})
+		},
+	}
+
+	indexers.AddIfNotPresentOrDie(
+		apiBindingInformer.Informer().GetIndexer(),
+		cache.Indexers{
+			indexers.APIBindingByBoundResourceUID: indexers.IndexAPIBindingByBoundResourceUID,
+		},
+	)
+
+	crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+
+			if !ok {
+				return false
+			}
+
+			return logicalcluster.From(crd) == apibinding.ShadowWorkspaceName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueCRD(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+		},
+	})
+
+	apiBindingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueFromAPIBinding(obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+// controller reconciles APIExports. It ensures an export's identity secret exists and is valid.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	getCRD                           func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error)
+	getAPIBindingsByBoundResourceUID func(name string) ([]interface{}, error)
+	listAPIBindings                  func() ([]*apisv1alpha1.APIBinding, error)
+	deleteCRD                        func(ctx context.Context, name string) error
+}
+
+// enqueueCRD enqueues a CRD.
+func (c *controller) enqueueCRD(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing CRD")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueFromAPIBinding(obj interface{}) {
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), binding)
+
+	for _, boundResource := range binding.Status.BoundResources {
```

Comment:
Normally, we'd have an index on the CRD informer to track CRDs by their binding, and queue all matching CRDs. I *think* using the list here in status is still correct, but it would be good to ACK that we've thought through the edge cases here to ensure that we can't miss a CRD if it's still related to the binding but is not listed in the bound resources array.

Reply:
Yes, the index will be updated first.

Yes, removing APIRS from an APIExport is possible.